### PR TITLE
ENH: Added test call from setup.py

### LINKED ISSUE #214: Add 'test' command to setup.py
The unit tests for cartopy can be run by calling `nosetests`. However, doing so will cause the the default matplotlib backend to be used. This default varies between platforms, e.g. under OSX it may be 'MacOSX', under linux it may be 'TkAgg'. The differences between these backends causes problems in the unit testing (see #211).

Modify `setup.py` so that one can run the units tests by entering   `python setup.py test`. Doing so should select the 'Agg' matplotlib backend and use the `KnownFailure` plugin. See below for a script that provides the basic functionality outside of `setup.py`:

``` python
#!/usr/bin/env python
#
# This allows running the cartopy tests from the command line: e.g.
# python tests.py -v -d

import matplotlib
matplotlib.use('agg')

import nose
from matplotlib.testing.noseclasses import KnownFailure


def run():
    nose.main(addplugins=[KnownFailure()])


if __name__ == '__main__':
    run()
```

### PR DESCRIPTION
**Pull request by cpelley**
Closes #214

Does not yet make use of multiprocessing as this causes an issue between tests currently (not enough time to sort this out for this PR).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -83,6 +83,52 @@ class MissingHeaderError(Exception):
...
+    
+    def finalize_options(self):
+        if self.stop:
+            print "Stopping tests after the first error or failure"
+
+    def run(self):
+        import nose
+        import matplotlib
+
+        matplotlib.use('agg')
```
Comment (QuLogic) [inline setup.py]: I think this should not be necessary due to #500?

File: setup.py [hunk 2]
```diff
@@ -83,6 +83,52 @@ class MissingHeaderError(Exception):
...
+        args = ['', None, '--with-doctest', '--nocapture', '--verbose']
+#                '--processes=%s' % n_processors,
+        if self.stop:
+            args.append('--stop')
+
+        result = True
+        for test in tests:
+            args[1] = test
+            print
+            print 'Running test discovery on {}.'.format(test)
```
Comment (QuLogic) [inline setup.py]: Should be `print()` now.

File: setup.py [hunk 3]
```diff
@@ -83,6 +83,52 @@ class MissingHeaderError(Exception):
     pass
 
 
+class TestRunner(Command):
+    """Run the Cartopy tests under nose"""
+    description = "run tests under nose"
+    user_options = [('stop', 'x', 'Stop running tests after the first '
+                     'error or failure'),]
+
```
Comment (QuLogic) [inline setup.py]: There's some trailing whitespace here.

File: setup.py [hunk 4]
```diff
@@ -83,6 +83,52 @@ class MissingHeaderError(Exception):
...
+    boolean_options = ['stop']
+
```
Comment (QuLogic) [inline setup.py]: There's some trailing whitespace here.

File: setup.py [hunk 5]
```diff
@@ -83,6 +83,52 @@ class MissingHeaderError(Exception):
...
+    """Run the Cartopy tests under nose"""
+    description = "run tests under nose"
+    user_options = [('stop', 'x', 'Stop running tests after the first '
+                     'error or failure'),]
+
+    boolean_options = ['stop']
+    
+    def initialize_options(self):
+        self.stop = False
+
```
Comment (QuLogic) [inline setup.py]: And here.

**Comment by QuLogic** [comment]
Does this actually work?
```
$ git clone --branch setup_test https://github.com/cpelley/cartopy.git cp
$ cd cp
$ python setup.py test
running test

Running test discovery on .../cp/lib/cartopy/tests.
Failure: ImportError (No module named _crs) ... ERROR

======================================================================
ERROR: Failure: ImportError (No module named _crs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File ".../cp/lib/cartopy/__init__.py", line 109, in <module>
    import cartopy.crs
  File ".../cp/lib/cartopy/crs.py", line 36, in <module>
    from cartopy._crs import CRS, Geocentric, Geodetic, Globe, PROJ4_RELEASE
ImportError: No module named _crs

----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
```

**Comment by pelson** [comment]
I think this PR could be made a whole lot easier now that @QuLogic has simplified the test running process and we are using setuptools. It might be worth taking a fresh look at how you would implement this given these changes.

I'll close the PR in the meantime - please feel free to re-open, or open another PR with the (hopefully) easier implementation.

**Comment by cpelley** [comment]
Thank you interest and chasing this up.  It was an age since I wrote this and no doubt needs numerous changes to reflect whats changed in Cartopy.

It was undertaken as we were putting out a release at the time (0.11.0) and were ensuring that issues under the 0.11.0 milestone were completed or associated with a new milestone.

I would still consider this low hanging fruit compared to other things at this time especially as its associated milestone is 0.13

I support the PR closure.

**Pull request closed without merging**